### PR TITLE
Improve wire ordering in tapering workflows

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Improves wire ordering in `_taper_pauli_sentence `

**Description of the Change:** Fixes an edge case related to Identities and simplifies some blocks using arithmetic. 

**Benefits:** 

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/fermi/fermionic.py [hunk 1]
```diff
@@ -721,7 +721,7 @@ def _to_string(fermi_op, of=False):
     pl_to_of_map = {"+": "^", "-": ""}
 
     if len(fermi_op) == 0:
-        return "I"
+        return "I" if not of else ""
```
Comment (obliviateandsurrender) [inline pennylane/fermi/fermionic.py:724]:
```suggestion
        return "I"
```

File: pennylane/qchem/tapering.py [hunk 2]
```diff
@@ -293,44 +293,39 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
     for ps in _split_pauli_sentence(ps_h, max_size=PAULI_SENTENCE_MEMORY_SPLITTING_SIZE):
         ts_ps += ps_u @ ps @ ps_u  # helps restrict the peak memory usage for u @ h @ u
 
-    wireset = ps_u.wires + ps_h.wires
+    wireset = ps_h.wires + ps_u.wires
```
Comment (austingmhuang) [inline pennylane/qchem/tapering.py:296]: Why is this needed?

Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:296]: Using the original Hamiltonian's wires first is better.

**Review by austingmhuang — COMMENTED** [review]
Nice work @obliviateandsurrender. Just to clarify, this is just to make tapering JAX-JIT compatible?

File: pennylane/qchem/tapering.py (hunk 2 above)
Comment (austingmhuang) [reply pennylane/qchem/tapering.py:296]: Can you provide an example of how this works?

**Comment by austingmhuang** [comment]
> @austingmhuang this is to fix the wire ordering issue with tapered Hamiltonians as raised by a user in the forum. Jax-jit support is additional.

Do we have a test case for this wire ordering issue?

**Comment by CatalinaAlbornoz** [comment]
Linking the [Forum thread](https://discuss.pennylane.ai/t/qubit-tapering-lih/7956/4) where the issue was first identified (for future reference).

File: pennylane/qchem/tapering.py [hunk 3]
```diff
@@ -293,44 +293,39 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
...
+        tapered_ham = qml.sum(*(qml.s_prod(coeff, op) for coeff, op in zip(coeffs, obs)))
+    else:
+        if qml.math.all(qml.math.abs(qml.math.imag(coeffs)) <= 1e-8):
+            coeffs = qml.math.real(coeffs)
+        tapered_ham = qml.simplify(0.0 * qml.Identity(wires=wires_ord) + qml.dot(coeffs, obs))
 
-        return tapered_ham + (0.0 * identity_op)
+    # If simplified Hamiltonian is missing wires, then add wires manually for consistency
+    if set(wires_ord) != tapered_ham.wires.toset():
+        return 0.0 * qml.Identity(wires=wires_ord) + tapered_ham
```
Comment (soranjh) [inline pennylane/qchem/tapering.py:328]: Not sure I understand this. Is this identity added to make sure there is a term with all wires defined in it? Does it mean that the Hamiltonian will have the term `0 * Identity`?

File: tests/qchem/test_tapering.py [hunk 4]
```diff
@@ -989,3 +989,44 @@ def test_split_pauli_sentence(ps_size, max_size):
...
+            np.array(
+                [[0.0, 0.0, 0.0], [0.0, 0.0, 1.4588684632]],
+                requires_grad=True,
+            ),
+            1,
+        ),
+    ],
+)
+def test_taper_jax_jit(symbols, geometry, charge):
+    r"""Test that an observable can be tapred within a jax-jit workflow."""
```
Comment (soranjh) [inline tests/qchem/test_tapering.py:1014]: How this test checks for wire ordering issues?

File: pennylane/qchem/tapering.py (hunk 3 above)
Comment (soranjh) [reply pennylane/qchem/tapering.py:328]: Might be good to have a look at how qubit_observable [fixes](https://github.com/PennyLaneAI/pennylane/blob/70a5700ec41b4c6b4af46e95bf451e3a55e76576/pennylane/qchem/observable_hf.py#L125) a similar issue.

Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:328]: Yes, this adds an `0 * Identity` with all the wires. This is supposed to catch some edge cases in tapering the Hartree-Fock, where the resulting simplified tapered observable acts on fewer qubits than expected. 

Regarding the solution with the `qubit_observable` - `qml.dot` returns to us an arithmetic operator, this would result in an overhead of obtaining the `pauli_rep` and converting it back to the operation and still wouldn't catch this particular edge case.

File: tests/qchem/test_tapering.py (hunk 4 above)
Comment (obliviateandsurrender) [reply tests/qchem/test_tapering.py:1032]: Added a test case that actually checks it.

File: tests/qchem/test_tapering.py [hunk 5]
```diff
@@ -989,3 +989,62 @@ def test_split_pauli_sentence(ps_size, max_size):
...
 
     assert sentence == qml.pauli.PauliSentence(split_sentence)
+
+
+@pytest.mark.parametrize(
+    ("symbols", "geometry"),
+    [(["Li", "H"], np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 3.13]]))],
+)
+def test_taper_wire_order(symbols, geometry):
+    r"""Test that an tapering workflow result in correct order of wires."""
```
Comment (soranjh) [inline tests/qchem/test_tapering.py:999]:
```suggestion
    r"""Test that a tapering workflow results in correct order of wires."""
```

File: pennylane/qchem/tapering.py (hunk 3 above)
Comment (soranjh) [reply pennylane/qchem/tapering.py:328]: Do the changes in this PR affect the results of the tapering demo and examples in other tapering functions?

Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:328]: Nope, they do not. Results would change only in the edge cases where a wire in the middle was tapered.

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -119,6 +119,9 @@
 
 * The `qml.clifford_t_decomposition` has been improved to use less gates when decomposing `qml.PhaseShift`.
   [(#6842)](https://github.com/PennyLaneAI/pennylane/pull/6842)
+ 
+* `qml.qchem.taper` now handles wire ordering for the tapered observables more robustly.
+  [(#6954)](https://github.com/PennyLaneAI/pennylane/pull/6954)
```
Comment (austingmhuang) [inline doc/releases/changelog-dev.md:124]: Maybe should be in bug fix section instead?

File: pennylane/qchem/tapering.py (hunk 3 above)
Comment (austingmhuang) [reply pennylane/qchem/tapering.py:328]: >This is supposed to catch some edge cases in tapering the Hartree-Fock, where the resulting simplified tapered observable acts on fewer qubits than expected.

What do you mean by this? As in the tapering resulted in not just a wire ordering problem but tapered more qubits than it's supposed to, leading to an incorrect observable?

Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:328]: As discussed offline - this particular case deals with tapering of the observable for the hf state. The tapered observable will still be correct but the reconstructed state from it will have a mismatch with the corresponding tapered Hamiltonian of that system.

File: pennylane/qchem/tapering.py [hunk 7]
```diff
@@ -293,44 +293,39 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
...
-                wire_map=wiremap_tap,
+        obs.append(
+            qml.pauli.PauliWord({wiremap_tap[wire]: pw[wire] for wire in wires_tap}).operation(
+                wire_order=wires_ord
             )
         )
 
-    c = qml.math.stack(qml.math.multiply(val * complex(1.0), list(ts_ps.values())))
+    interface = qml.math.get_deep_interface(list(ps_h.values()))
+    coeffs = qml.math.multiply(val, qml.math.array(list(ts_ps.values()), like=interface))
```
Comment (austingmhuang) [inline pennylane/qchem/tapering.py:317]: val is just an array of 1s? Why do we need to multiply by 1s?

Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:317]: Vals is an array of `[+1, -1, +1j, -1j]`, essentially the phases we pick up for each qubit we are tapering. These get multiplied by the coefficients to retain the information stored on those qubits.

File: doc/releases/changelog-dev.md (hunk 6 above)
Comment (obliviateandsurrender) [reply doc/releases/changelog-dev.md:124]: Won't call it a bug myself, but can shift it there.

File: pennylane/qchem/tapering.py [hunk 8]
```diff
@@ -293,44 +293,39 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
...
-        )
+    if interface == "jax" and qml.math.is_abstract(coeffs):
+        tapered_ham = qml.sum(*(qml.s_prod(coeff, op) for coeff, op in zip(coeffs, obs)))
+    else:
+        if qml.math.all(qml.math.abs(qml.math.imag(coeffs)) <= 1e-8):
+            coeffs = qml.math.real(coeffs)
+        tapered_ham = qml.simplify(0.0 * qml.Identity(wires=wires_ord) + qml.dot(coeffs, obs))
 
-        return tapered_ham + (0.0 * identity_op)
+    # If simplified Hamiltonian is missing wires, then add wires manually for consistency
```
Comment (austingmhuang) [inline pennylane/qchem/tapering.py:326]: Minor: this comment could use a bit more context, i.e. answer the question "Why would the simplified Hamiltonian miss wires?"

File: pennylane/qchem/tapering.py [hunk 9]
```diff
@@ -293,44 +293,39 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
...
-    wiremap_tap = dict(zip(wires_tap, range(len(wires_tap) + 1)))
-
-    for i, pw_coeff in enumerate(ts_ps.items()):
-        pw, _ = pw_coeff
+    wires_tap = [i for i in wiremap.keys() if i not in paulix_wires]
+    wires_ord = list(range(len(wires_tap)))
+    wiremap_tap = dict(zip(wires_tap, wires_ord))
 
+    obs, val = [], qml.math.ones(len(ts_ps))
+    for i, (pw, _) in enumerate(ts_ps.items()):
```
Comment (austingmhuang) [inline pennylane/qchem/tapering.py:305]: Minor: Could use `ts_ps_values()` instead? Then line 317 you can also replace it with a variable.

File: pennylane/qchem/tapering.py [hunk 10]
```diff
@@ -293,44 +293,39 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
...
-    val = np.ones(len(ts_ps))
-
-    wires_tap = [i for i in ts_ps.wires if i not in paulix_wires]
-    wiremap_tap = dict(zip(wires_tap, range(len(wires_tap) + 1)))
-
-    for i, pw_coeff in enumerate(ts_ps.items()):
-        pw, _ = pw_coeff
+    wires_tap = [i for i in wiremap.keys() if i not in paulix_wires]
+    wires_ord = list(range(len(wires_tap)))
+    wiremap_tap = dict(zip(wires_tap, wires_ord))
```
Comment (austingmhuang) [inline pennylane/qchem/tapering.py:302]: Minor: could do this in one for loop.

**Review by austingmhuang — APPROVED** [review]
Thanks @obliviateandsurrender.

File: pennylane/qchem/tapering.py (hunk 10 above)
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:302]: I use two of these variables later on in the code, so would prefer to leave it this way.

File: pennylane/qchem/tapering.py (hunk 9 above)
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:305]: I think you meant `keys`, can do that.

File: pennylane/qchem/tapering.py (hunk 10 above)
Comment (austingmhuang) [reply pennylane/qchem/tapering.py:302]:
```suggestion
    wires_tap, wires_ord, wiremap_tap, length = [], [], {}, 0
    for i in wiremap.keys():
        if i not in paulix_wires:
            wires_tap.append(i)
            wires_ord.append(length)
            length += 1
            wiremap_tap[i] = length
```

Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:302]: overkill for these variables.

Comment (austingmhuang) [reply pennylane/qchem/tapering.py:302]: Let's review this some more. I think that you don't need these variables.

When you do
`wiremap_tap[wire]: pw[wire] for wire in wires_tap`, what is the expected value?

From what I can tell, you will always get a dictionary that has keys ascending from 0 to N. Do we _need_ this dictionary?

File: pennylane/qchem/tapering.py [hunk 11]
```diff
@@ -293,44 +293,40 @@ def _taper_pauli_sentence(ps_h, generators, paulixops, paulix_sector):
...
             if pw[w] == "X":
                 val[i] *= paulix_sector[idx]
 
-        o.append(
-            qml.pauli.string_to_pauli_word(
-                "".join([pw[wiremap[i]] for i in wires_tap]),
-                wire_map=wiremap_tap,
+        obs.append(
+            qml.pauli.PauliWord({wiremap_tap[wire]: pw[wire] for wire in wires_tap}).operation(
+                wire_order=wires_ord
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/qchem/tapering.py:312]: Technically I can directly generate everything straight from `wiremap` and `paulix_wires` defined in lines 297-298, without needing to define anything. I use `wires_ord` further down for Identities, so wouldn't want to redefine it again. Can consider removing the `wiremap_tap` after some testing.
